# Question: 

Discussion about the imbalanced nature of the data and how you want to address it?

In [1]:
import pandas as pd

In [9]:
imdb_milestone2 = pd.read_csv(r"C:\Users\cheriexu\Downloads\CS109B\imdb_cluster_result.csv")

In [11]:
imdb_milestone2.head()

X  certificates_R  certificates_PG  art.direction_1  \
0    100               1                0         0.322495   
1  10001               0                1         0.027673   
2  10002               1                0         0.212394   
3  10003               0                1         0.019767   
4  10004               1                0         0.104764   

   assistant.director_1  casting.director_1  cinematographer_1  \
0              0.042103            0.010279           0.046254   
1              0.093694            0.319431           0.165250   
2              0.024906            0.006523           0.016308   
3              0.024906            0.014133           0.025301   
4              0.355406            0.017395           0.049812   

   costume.department_1  costume.designer_1  countries_1  \
0              0.307966            0.028662     0.089642   
1              0.307966            0.172860     0.014232   
2              0.012453            0.005535     0.089642   
3              0.126112            0.007116     0.536766   
4              0.126112            0.040028     0.536766   

             ...             Action  Documentary  Musical  History  Family  \
0            ...                  0            0        0        0       0   
1            ...                  0            0        0        1       0   
2            ...                  0            0        0        0       0   
3            ...                  1            0        0        0       0   
4            ...                  0            0        0        0       0   

   Fantasy  Sport  Biography  cluster_response               genres_comb  
0        0      0          0                 5  Horror, Thriller, Action  
1        0      0          0                 4  Romance, Comedy, Fantasy  
2        0      0          0                 1   Horror, Thriller, Drama  
3        0      0          0                 1   Horror, Thriller, Drama  
4        1      0          0                 1   Horror, Thriller, Drama  

[5 rows x 58 columns]

In our data set, we use "cluster_response" (i.e genre_comb) as our response variable. 

In [15]:
imdb_milestone2.ix[:,'cluster_response'].value_counts()

1    3621
5    2641
4    1953
2    1850
3      53
Name: cluster_response, dtype: int64

By above frequency table, we can tell classes are not balanced in our processed data set. so we would like to do the following data sampling to deal with imbalanced sitaution.

## Data Sampling

### Reason for sampling :

Imbalanced classes: most classification algorithms work relatively well on balanced classes where the number of samples of each class is similar. However, when some classes have outnumbered samples than others, there will be a difficulty solving classificiation problems with standard algorithms. Since these algorithms trend to minimize quantities such as error rate, they will more likely to predict towards the majority class. In order to minimize the problem caused by inbalanced classes, re-sampling the dataset as to offset the imbalance situation will be a pausible choice. 


### Possible Re-sampling Methods

There are three major types of re-sampling techniques:

1. Under-sampling (the majority classes).

2. Over-sampling (the minority classes).

3. Combination of under-sampling & over-sampling

### Under-sampling:

**1. Randomly sample the majority class(es) samples with replacement:**

    Set up a threshold to divide classes into minority classes and majority classes. Then randomly sample the movies in each majority class until the number of movies reach the average number of movies in the minority classes. When we sample with replacement, the two sample values are independent. 
   

In [19]:
#possible useful package:
from imblearn.under_sampling import RandomUnderSampler

**2. Cluster-based undersampling:**

    cluster the majority classes samples to k clusters. Combine each cluster of majority class sample with all the minoirty class sample and make K combination of datasets. Classify all the datasets and choose the datasets out of K combined 
    datasets that gives the highest accuracy, regard it as the final training sample for building the decision support.

In [20]:
#possible useful package & method:
from imblearn.under_sampling import ClusterCentroids 

paper reference: http://www.iaeng.org/publication/WCE2013/WCE2013_pp1480-1485.pdf

**3. Nearest Neighbour based undersampling:**

(1) condensed nearest neighbor rule (CNN):

    select a sub-set A from the original data T which is consistent with T in the sense that A classifies T correctly with the one-nearest neighbor method. In specific, it starts from a set of data including all examples of the class of interest C and by removing examples and one randomly selected example from each class in rest of data O (= T - C), until without misclassifications.
    
    Drawbacks: extremely sensitive to noisy and many of them will be added to the training set. 
    
(2) Edited nearest neighbor rule (ENN):

    ENN removes examples whose class label differs from the majority class of the three nearest neighbors
    
(3) Neighborhood Cleaning rule:

    Build on CNN and ENN, and identification and possible removal of outliers. 
    1. Split data T into the class of interest C and the rest of data O.
    2. Identify noisy data A1 in O with the edited nearest neighbor rule.
    3. For each class Ci in O, three nearest neighbors that misclassify examples are inserted into set A2. To avoid excessive reduction of small classes, only examples from classes larger or equal than 0.5 · | C | are considered while forming A2
    4. Reduced data S = T - ( A1 ∪ A2 ) 

In [21]:
#possible useful package & method:
from imblearn.under_sampling import NeighbourhoodCleaningRule
from imblearn.under_sampling import EditedNearestNeighbours 
from imblearn.under_sampling import CondensedNearestNeighbour

paper reference: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.577.4704&rep=rep1&type=pdf

One possible issue of under-sampling is that if we have each combination of labels as a class, then there are possible many minority classes with few movies. Under-sampling majority classes may give us much fewer observations than the original data set. 

### Over-sampling:

**1. Randomly sample the minority class(es) samples with replacement:**

    Set up a threshold to divide classes into minority classes and majority classes. Then randomly sample the movies in each minority class until the number of movies reach the average number of movies in the majority classes.


In [22]:
#possible useful package & method:
from imblearn.over_sampling import RandomOverSampler

**2. Synthetic Minority Over-sampling Technique - SMOTE:**

The minority class is over-sampled by taking each minority class sample and introducing synthetic examples along the line
segments joining any/all of the k minority class nearest neighbors. 

For instance, if the amount of over-sampling needed is 200%, only two neighbors from the five (if k=5) nearest neighbors
are chosen and one sample is generated in the direction of each. Synthetic samples are generated in the following way: Take the difference between the feature vector (sample) under consideration and its nearest neighbor. Multiply this difference by a random number between 0 and 1, and add it to the feature vector under consideration. This causes the selection of a random point along the line segment between two specific features. This approach effectively forces the decision region of the minority class to become more general.

In [24]:
#possible useful package & method:
from imblearn.over_sampling import SMOTE 

paper reference: https://www.jair.org/media/953/live-953-2037-jair.pdf

**3. Adaptive synthetic sampling approach for imbalanced learning -ADASYN**

It uses a density distribution as a criterion to automatically decide the number of synthetic samples that need to be generated for each minority data example. The density distribution is a measurement of the distribution of weights for different minority class examples according to their level of difficulty in learning. It provides different weights for different minority examples to compensate for the skewed distributions.

Difference from SMOTE:  The resulting dataset post ADASYN will not only provide a balanced representation of the data distribution, but it will also force the learning algorithm to focus on those difficult to learn examples, where SMOTE has equal numbers of synthetic samples are generated for each minority data example.


In [25]:
#possible useful package & method:
from imblearn.over_sampling import ADASYN 

paper reference: http://ieeexplore.ieee.org/document/4633969/?tp=&arnumber=4633969&url=http:%2F%2Fieeexplore.ieee.org%2Fxpls%2Fabs_all.jsp%3Farnumber%3D4633969

One concern of over-sampling is that we already have large amount of data set, over-sampling will increase samples, which causes longer run-time.

### Combination of under-sampling & over-sampling

**1. First oversampling then undersampling: SMOTE (introduced in under-sampling) + Tomek links:**

Frequently, class clusters are not well defined since some majority class examples might be invading the minority class space. The opposite can also be true, since interpolating minority class examples can expand the minority class clusters, introducing artificial minority class examples too deeply in the majority class space. Inducing a classifier under such a situation can lead to overfitting. In order to better-defined class clusters, we can apply Tomek links to the over-sampled training set as a data cleaning method. 

**2. First undersampling then oversampling: SMOTE (introduced in under-sampling) + ENN (introduced in over-sampling):**

Similar to above, except ENN tends to remove more examples than the Tomek links does, so it is expected that it will provide a more in depth data cleaning.Thus, any example that is misclassified by its three nearest neighbors is removed from the training set.

paper reference: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.58.7757&rep=rep1&type=pdf

## Summary:

Three are benefits and drawbacks of each sampling methods. 

**Under-sampling**:

Benefit:reduce sample size (lower runtime); more types of sampling methods.

Drawbacks: much fewer observations than the original data set.

**Over-sampling**

Benefit:enough data set for each class.

Drawbacks: even longer runtime.

**Over-sampling and Under-sampling:**

Benefit:avoid over-fitting by data cleaning.

Drawbacks: not suitable for multiple classes.


We can choose sampling method based on our final decision of the response variable, since some methods are not appliable to multi-class classification problems. However, if we would like to do the classificiation 1 by 1 (for example, Action movie vs Not Action movie), we can still use all of these methods. Right now, we would like to regard the response variable as a multi-class (each combination of labels/cluster as a class), we can only use sampling methods that are suitable for multiple classes.

**Note: methods are applicable to multiple classes:**

Under-sampling: (1) Random under-sampling (2) Condensed Nearest Neighbour (3) Edited nearest neighbor rule (4) Neighbourhood CleaningRule

Over-sampling: (1) Random Over-sampling: Supports multiple classes. (2) SOMTE: It does not support multiple classes automatically, but can be called multiple times.





### Other  Useful References:

Python class: imbalanced-learn: http://contrib.scikit-learn.org/imbalanced-learn/index.html

Systematic Introducation of imbalanced classes (including evalution & after sampling algorithm level):  https://svds.com/learning-imbalanced-classes/

# Questions: 
How do you sample your data, how many samples, and why?

Large data set: IMDB and TMDB have large amounts movie data information available. Using all of them for the later model building will be computational challenge. Also, as we discussed earlier, there is an imbalanced class distribution in our dataset, so we would like to re-sampling data to solve the issue. Possible solution is using under-sampling to reduce the number of movies and resolve the imbalanced issue. 


### Data example for re-sampling

In [38]:
imdb_milestone2.shape[0]

10118

In milestone2, we didn't use all the movies due to longer than expected processing time to loading data, we use 10118 movies as example. In the later milestones, we will use all availble data set (after filtering movies does not meet our need). We estimated we will have more than 200,000 movies in the whole data set, so we would like to under-sampling larger classes to achieve a balance.

In [26]:
from imblearn.under_sampling import RandomUnderSampler
from collections import Counter

In [21]:
X = imdb_milestone2.ix[:,"X":"rating"]

In [23]:
y = imdb_milestone2.ix[:, "cluster_response"]

In [28]:
print('Resampled dataset shape {}'.format(Counter(y)))

Resampled dataset shape Counter({1: 3621, 5: 2641, 4: 1953, 2: 1850, 3: 53})


In [24]:
rus = RandomUnderSampler(random_state=42)
X_res, y_res = rus.fit_sample(X, y)

In [27]:
print('Resampled dataset shape {}'.format(Counter(y_res)))

Resampled dataset shape Counter({1: 53, 2: 53, 3: 53, 4: 53, 5: 53})


In [29]:
#If we don't want the data set reduce too much, we can set a ratio
rus = RandomUnderSampler(random_state=42, ratio = 0.4)
X_res, y_res = rus.fit_sample(X, y)
print('Resampled dataset shape {}'.format(Counter(y_res)))

Resampled dataset shape Counter({1: 132, 2: 132, 4: 132, 5: 132, 3: 53})


In [30]:
from imblearn.under_sampling import NeighbourhoodCleaningRule
from imblearn.under_sampling import EditedNearestNeighbours 
from imblearn.under_sampling import CondensedNearestNeighbour

Besides random sampling, we can also use neighborhood based sampling methods, especially Neighborhood Cleaning rul. It combines the strength of Edited Nearest Neighbours and Condensed Nearest Neighbour, as we discussed ealier. However, it may gives us still an imbalanced classes if we set n_neighbors small. It also will decrease the number of observations in the minority class.

In [34]:
ncr = NeighbourhoodCleaningRule(random_state=42,n_neighbors = 5)
X_res, y_res = ncr.fit_sample(X, y)
print('Resampled dataset shape {}'.format(Counter(y_res)))

Resampled dataset shape Counter({1: 382, 3: 53, 5: 50, 2: 35, 4: 34})


We would like to use either Random undersampling or CNN undersampling for the whole data set in the later milestones, based on how small the minority class is. Since CNN might drop the observation in the minority class, if we have very few observations in the minority class, we would like to use random sampling instead. 